File: server/torbox.mjs
export const ALLOWED_FETCH_HOSTS = [
  'v3-cinemeta.strem.io',
  'caching.stremio.net',
  'torrentio.strem.fun',
  'comet.elfhosted.com',
  'mediafusion.elfhosted.com',
  'api.torbox.app',
]

const VIDEO_FILENAME_PATTERN = /\.(mkv|mp4|avi|mov|webm|m4v)$/i
const TORRENT_APPEAR_TIMEOUT_MS = 120_000
const TORRENT_READY_TIMEOUT_MS = 180_000
const TORRENT_POLL_INTERVAL_MS = 2_000
const DOWNLOAD_LINK_ATTEMPTS = 5

function bodyDetail(body) {
  if (!body || typeof body !== 'object') return ''
  return body.detail || body.error || body.message || body.msg || ''
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export function isQueuedTorrentDetail(detail) {
  const message = String(detail || '').toLowerCase()
  return message.includes('queued') || message.includes('notification when it is processed')
}

export function isRetriableTorboxError(message) {
  const text = String(message || '').toLowerCase()
  return (
    text.includes('there was an error processing your request')
    || text.includes('please try again later')
    || text.includes('internal server error')
    || text.includes('bad gateway')
    || text.includes('service unavailable')
    || text.includes('gateway timeout')
    || /^5\d\d/.test(text)
  )
}

export function extractTorrentId(body) {
  if (!body || typeof body !== 'object') return null
  const candidates = [
    body?.data?.torrent_id,
    body?.data?.id,
    body?.torrent_id,
    body?.id,
  ]
  for (const candidate of candidates) {
    const parsed = Number(candidate)
    if (Number.isFinite(parsed) && parsed > 0) return parsed
  }
  return null
}

export function isTorrentCached(torrent) {
  if (!torrent || typeof torrent !== 'object') return false
  if (torrent.cached === true) return true
  const state = String(torrent.download_state || torrent.state || '').toLowerCase()
  return state.includes('cached') || state.includes('completed') || state.includes('complete')
}

function torrentFiles(torrent) {
  return torrent?.files || torrent?.file || torrent?.filelist || torrent?.contents || []
}

export function hasSelectableVideoFile(torrent) {
  const files = torrentFiles(torrent)
  if (!Array.isArray(files)) return false
  return files.some((file) => isVideoFilename(fileName(file)) && numericFileId(file) !== null)
}

async function torboxFetch(url, options = {}) {
  const response = await fetch(url, options)
  const text = await response.text()
  let body = null
  try {
    body = text ? JSON.parse(text) : null
  } catch {
    body = text
  }
  const detail = bodyDetail(body)
  if (!response.ok || body?.success === false) {
    throw new Error(detail || (typeof body === 'string' ? body : `${response.status} ${response.statusText}`))
  }
  return body
}

async function fetchJson(url, options = {}) {
  return torboxFetch(url, options)
}

export function extractInfoHash(value) {
  const match = String(value || '').match(/btih:([^&?/]+)/i)
  return match?.[1]
}

function fileName(file) {
  return String(file?.name || file?.short_name || file?.path || file?.filename || '')
}

function fileSize(file) {
  return Number(file?.size || file?.bytes || file?.length || 0)
}

function numericFileId(file) {
  const value = file?.id ?? file?.file_id ?? file?.fileId ?? file?.idx ?? file?.index
  const parsed = Number(value)
  return Number.isFinite(parsed) ? parsed : null
}

export function isVideoFilename(name) {
  return VIDEO_FILENAME_PATTERN.test(String(name || ''))
}

export function chooseFileId(torrent, fileIdx, filename) {
  const files = torrentFiles(torrent)
  if (!Array.isArray(files)) return 0

  const indexed = Number.isFinite(fileIdx) ? files[Number(fileIdx)] : null
  if (indexed && isVideoFilename(fileName(indexed)) && numericFileId(indexed) !== null) {
    return numericFileId(indexed) ?? 0
  }

  if (filename) {
    const needle = String(filename).toLowerCase()
    const matched = files.find((file) => {
      const name = fileName(file).toLowerCase()
      return isVideoFilename(name) && (needle.includes(name) || name.includes(needle))
    })
    if (matched && numericFileId(matched) !== null) return numericFileId(matched) ?? 0
  }

  const bestVideo = files
    .filter((file) => isVideoFilename(fileName(file)) && numericFileId(file) !== null)
    .sort((a, b) => fileSize(b) - fileSize(a))[0]

  return numericFileId(bestVideo) ?? numericFileId(indexed) ?? numericFileId(files[0]) ?? 0
}

function isAllowedHost(hostname) {
  const lowerHost = String(hostname || '').toLowerCase()
  if (lowerHost === 'strem.io' || lowerHost.endsWith('.strem.io')) return true
  return ALLOWED_FETCH_HOSTS.includes(lowerHost)
}

export function normalizeAllowedFetchJsonUrl(value) {
  const url = new URL(String(value || '').trim())
  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    throw new Error('JSON fetches only support HTTP and HTTPS URLs.')
  }
  if (!isAllowedHost(url.hostname)) {
    throw new Error(`Host ${url.hostname} is not in the JSON fetch allowlist.`)
  }
  return url.toString()
}

async function findTorrentByHash(token, hash, bypassCache = false) {
  const url = new URL('https://api.torbox.app/v1/api/torrents/mylist')
  if (bypassCache) url.searchParams.set('bypass_cache', 'true')
  url.searchParams.set('limit', '1000')
  const body = await fetchJson(url.toString(), {
    headers: { Authorization: `Bearer ${token}` },
  })
  const torrents = Array.isArray(body?.data) ? body.data : Array.isArray(body) ? body : []
  const wanted = hash.toLowerCase()
  const found = torrents.find((torrent) => {
    const candidates = [torrent.hash, torrent.info_hash, torrent.infoHash, torrent.torrent_hash]
    return candidates.some((candidate) => String(candidate || '').toLowerCase() === wanted)
  })
  const id = found?.id ?? found?.torrent_id ?? found?.torrentId
  const parsed = Number(id)
  return Number.isFinite(parsed) && parsed > 0 ? parsed : null
}

async function waitForTorrentByHash(token, hash, timeoutMs = TORRENT_APPEAR_TIMEOUT_MS) {
  const deadline = Date.now() + timeoutMs
  while (Date.now() < deadline) {
    const id = await findTorrentByHash(token, hash, true)
    if (id) return id
    await sleep(TORRENT_POLL_INTERVAL_MS)
  }
  return null
}

async function createTorrent(token, hash, name) {
  const form = new FormData()
  form.set('magnet', `magnet:?xt=urn:btih:${hash}`)
  form.set('seed', '1')
  form.set('allow_zip', 'false')
  form.set('as_queued', 'false')
  if (name) form.set('name', name)
  const body = await fetchJson('https://api.torbox.app/v1/api/torrents/createtorrent', {
    method: 'POST',
    headers: { Authorization: `Bearer ${token}` },
    body: form,
  })
  let id = extractTorrentId(body)
  if (!id) id = await waitForTorrentByHash(token, hash)
  if (!id) {
    const detail = bodyDetail(body)
    if (isQueuedTorrentDetail(detail)) {
      throw new Error('Torbox queued this torrent but it did not appear in your library in time. Try again shortly.')
    }
    throw new Error(detail || 'Torbox did not return a torrent id.')
  }
  return id
}

async function getTorrent(token, torrentId, bypassCache = false) {
  const url = new URL('https://api.torbox.app/v1/api/torrents/mylist')
  url.searchParams.set('id', String(torrentId))
  if (bypassCache) url.searchParams.set('bypass_cache', 'true')
  const body = await fetchJson(url.toString(), {
    headers: { Authorization: `Bearer ${token}` },
  })
  return Array.isArray(body?.data) ? body.data[0] : body?.data || body
}

async function waitForTorrentReady(token, torrentId) {
  const deadline = Date.now() + TORRENT_READY_TIMEOUT_MS
  let lastState = ''
  while (Date.now() < deadline) {
    const torrent = await getTorrent(token, torrentId, true)
    lastState = String(torrent?.download_state || torrent?.state || '')
    const state = lastState.toLowerCase()
    if (state.includes('error') || state.includes('failed')) {
      throw new Error('Torbox reported an error caching this torrent. Try another result.')
    }
    if (isTorrentCached(torrent) && hasSelectableVideoFile(torrent)) return torrent
    if (isTorrentCached(torrent) && torrentFiles(torrent).length > 0) return torrent
    await sleep(TORRENT_POLL_INTERVAL_MS)
  }
  throw new Error(
    lastState
      ? `Torbox is still preparing this torrent (${lastState}). Wait a moment and try again.`
      : 'Torbox is still preparing this torrent. Wait a moment and try again.',
  )
}

function extractDownloadUrl(body) {
  if (typeof body?.data === 'string') return body.data
  const data = body?.data && typeof body.data === 'object' ? body.data : body
  for (const key of ['download_url', 'downloadUrl', 'url', 'link', 'download', 'redirect']) {
    if (typeof data?.[key] === 'string') return data[key]
  }
  throw new Error(bodyDetail(body) || 'Torbox did not return a direct download URL.')
}

async function requestDownloadLinkOnce(token, torrentId, fileId) {
  const url = new URL('https://api.torbox.app/v1/api/torrents/requestdl')
  url.searchParams.set('token', token)
  url.searchParams.set('torrent_id', String(torrentId))
  url.searchParams.set('file_id', String(fileId || 0))
  url.searchParams.set('zip_link', 'false')
  url.searchParams.set('redirect', 'false')
  url.searchParams.set('append_name', 'false')
  return extractDownloadUrl(await fetchJson(url.toString()))
}

async function requestDownloadLink(token, torrentId, fileId) {
  let lastError = null
  for (let attempt = 1; attempt <= DOWNLOAD_LINK_ATTEMPTS; attempt += 1) {
    try {
      return await requestDownloadLinkOnce(token, torrentId, fileId)
    } catch (error) {
      lastError = error
      const message = error instanceof Error ? error.message : String(error)
      if (!isRetriableTorboxError(message) || attempt === DOWNLOAD_LINK_ATTEMPTS) throw error
      await sleep(attempt * 1500)
    }
  }
  throw lastError instanceof Error ? lastError : new Error('Torbox did not return a direct download URL.')
}

function stringField(body, keys) {
  for (const key of keys) {
    const value = body?.[key]
    if (typeof value === 'string' && value.trim()) return value
  }
  return undefined
}

function numberField(body, keys) {
  for (const key of keys) {
    const value = body?.[key]
    if (typeof value === 'number' && Number.isFinite(value)) return value
  }
  return undefined
}

function planLabel(planId) {
  if (planId === 0) return 'Free'
  if (planId === 1) return 'Essential'
  if (planId === 2) return 'Pro'
  if (planId === 3) return 'Standard'
  return `Plan ${planId}`
}

export async function fetchTorboxAccount(token) {
  const apiKey = String(token || '').trim()
  if (!apiKey) throw new Error('Add your Torbox API key to load account details.')

  const userBody = await fetchJson('https://api.torbox.app/v1/api/user/me', {
    headers: { Authorization: `Bearer ${apiKey}` },
  })
  const user = userBody?.data && typeof userBody.data === 'object' ? userBody.data : userBody
  const planId = numberField(user, ['plan', 'plan_id', 'planId'])
  const premium = typeof user?.premium === 'boolean'
    ? user.premium
    : typeof user?.is_premium === 'boolean'
      ? user.is_premium
      : (planId ?? 0) > 0

  let torrents = []
  try {
    const torrentBody = await fetchJson('https://api.torbox.app/v1/api/torrents/mylist', {
      headers: { Authorization: `Bearer ${apiKey}` },
    })
    torrents = Array.isArray(torrentBody?.data) ? torrentBody.data : Array.isArray(torrentBody) ? torrentBody : []
  } catch {
    torrents = []
  }

  let activeTorrents = 0
  let cachedTorrents = 0
  for (const entry of torrents) {
    const state = String(entry?.download_state || entry?.state || '').toLowerCase()
    if (state.includes('cached') || entry?.cached === true) cachedTorrents += 1
    if (state && !['completed', 'complete', 'cached', 'error'].some((item) => state.includes(item))) {
      activeTorrents += 1
    }
  }

  return {
    email: stringField(user, ['email', 'user_email']),
    plan: stringField(user, ['plan_name', 'planName']) || (planId !== undefined ? planLabel(planId) : undefined),
    planId,
    premium,
    expiresAt: stringField(user, ['premium_expires_at', 'expires_at', 'expiration']),
    totalTorrents: torrents.length,
    activeTorrents,
    cachedTorrents,
  }
}

export async function resolveTorboxStream({ token, infoHash, fileIdx, filename, directUrl }) {
  if (directUrl && /^https?:\/\//.test(directUrl)) return directUrl
  const hash = infoHash || extractInfoHash(directUrl)
  if (!hash) throw new Error('This stream does not expose an info hash or playable URL.')
  if (!String(token || '').trim()) throw new Error('Add your Torbox API key before resolving Torbox results.')
  const cleanToken = token.trim()
  const existingTorrentId = await findTorrentByHash(cleanToken, hash, true)
  const torrentId = existingTorrentId || (await createTorrent(cleanToken, hash, filename))
  const torrent = await waitForTorrentReady(cleanToken, torrentId)
  const fileId = chooseFileId(torrent, Number(fileIdx), filename)
  return requestDownloadLink(cleanToken, torrentId, fileId)
}

File: src-tauri/src/torbox.rs
const ALLOWED_FETCH_HOSTS: &[&str] = &[
    "v3-cinemeta.strem.io",
    "torrentio.strem.fun",
    "comet.elfhosted.com",
    "mediafusion.elfhosted.com",
    "api.torbox.app",
];

fn is_allowed_fetch_host(hostname: &str) -> bool {
    let host = hostname.to_ascii_lowercase();
    if host == "strem.io" || host.ends_with(".strem.io") {
        return true;
    }
    ALLOWED_FETCH_HOSTS.contains(&host.as_str())
}

#[tauri::command]
pub(crate) async fn fetch_json(url: String) -> Result<serde_json::Value, String> {
    let parsed = reqwest::Url::parse(&url).map_err(|error| error.to_string())?;

    if parsed.scheme() != "https" && parsed.scheme() != "http" {
        return Err("Only HTTP and HTTPS URLs are supported".into());
    }
    let Some(hostname) = parsed.host_str() else {
        return Err("URL host is required".into());
    };
    if !is_allowed_fetch_host(hostname) {
        return Err(format!(
            "Host {hostname} is not in the JSON fetch allowlist."
        ));
    }

    let response = reqwest::Client::new()
        .get(parsed)
        .header(reqwest::header::ACCEPT, "application/json")
        .send()
        .await
        .map_err(|error| error.to_string())?;

    let status = response.status();
    if !status.is_success() {
        return Err(format!("Request failed with {}", status));
    }

    response
        .json::<serde_json::Value>()
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub(crate) async fn resolve_torbox_stream(
    token: String,
    info_hash: Option<String>,
    file_idx: Option<i64>,
    filename: Option<String>,
    direct_url: Option<String>,
) -> Result<String, String> {
    if let Some(url) = direct_url.as_deref() {
        if url.starts_with("http://") || url.starts_with("https://") {
            return Ok(url.to_string());
        }
    }

    let hash = info_hash
        .or_else(|| direct_url.as_deref().and_then(extract_info_hash))
        .ok_or_else(|| "This stream does not expose an info hash or playable URL.".to_string())?;

    let client = reqwest::Client::new();
    let token = token.trim().to_string();
    if token.is_empty() {
        return Err("Add your Torbox API key before streaming Torbox results.".into());
    }

    let torrent_id = match find_torrent_by_hash(&client, &token, &hash).await? {
        Some(id) => id,
        None => create_torrent(&client, &token, &hash, filename.as_deref()).await?,
    };

    let torrent = get_torrent(&client, &token, torrent_id).await?;
    let file_id = choose_file_id(&torrent, file_idx, filename.as_deref()).unwrap_or(0);
    request_download_link(&client, &token, torrent_id, file_id).await
}

pub(crate) fn extract_info_hash(value: &str) -> Option<String> {
    let lower = value.to_lowercase();
    let marker = "btih:";
    let start = lower.find(marker)? + marker.len();
    let rest = &value[start..];
    let hash = rest.split(['&', '?', '/']).next()?.trim();

    if hash.is_empty() {
        None
    } else {
        Some(hash.to_string())
    }
}

async fn find_torrent_by_hash(
    client: &reqwest::Client,
    token: &str,
    info_hash: &str,
) -> Result<Option<i64>, String> {
    let response = client
        .get("https://api.torbox.app/v1/api/torrents/mylist")
        .bearer_auth(token)
        .query(&[("bypass_cache", "true"), ("limit", "1000")])
        .send()
        .await
        .map_err(|error| error.to_string())?;

    let body = json_response(response, "Torbox torrent details").await?;
    let Some(items) = body.get("data").and_then(|data| data.as_array()) else {
        return Ok(None);
    };

    let needle = info_hash.to_lowercase();
    Ok(items.iter().find_map(|item| {
        let hash = item
            .get("hash")
            .or_else(|| item.get("info_hash"))
            .and_then(|value| value.as_str())?
            .to_lowercase();

        if hash == needle {
            numeric_id(item)
        } else {
            None
        }
    }))
}

async fn create_torrent(
    client: &reqwest::Client,
    token: &str,
    info_hash: &str,
    name: Option<&str>,
) -> Result<i64, String> {
    let magnet = format!("magnet:?xt=urn:btih:{info_hash}");
    let mut form = reqwest::multipart::Form::new()
        .text("magnet", magnet)
        .text("seed", "1")
        .text("allow_zip", "false")
        .text("as_queued", "false");

    if let Some(name) = name {
        form = form.text("name", name.to_string());
    }

    let response = client
        .post("https://api.torbox.app/v1/api/torrents/createtorrent")
        .bearer_auth(token)
        .multipart(form)
        .send()
        .await
        .map_err(|error| error.to_string())?;

    let status = response.status();
    let body = json_response(response, "Torbox download request").await?;

    if !status.is_success() || body.get("success").and_then(|value| value.as_bool()) == Some(false)
    {
        return Err(body_detail(&body)
            .unwrap_or_else(|| format!("Torbox create torrent failed with {status}")));
    }

    body.get("data")
        .and_then(|data| data.get("torrent_id").or_else(|| data.get("id")))
        .and_then(json_to_i64)
        .or_else(|| body.get("torrent_id").and_then(json_to_i64))
        .or_else(|| body.get("id").and_then(json_to_i64))
        .ok_or_else(|| "Torbox did not return a torrent ID.".to_string())
}

async fn json_response(
    response: reqwest::Response,
    context: &str,
) -> Result<serde_json::Value, String> {
    let status = response.status();
    let text = response.text().await.map_err(|error| error.to_string())?;
    serde_json::from_str::<serde_json::Value>(&text).map_err(|error| {
        let preview = text.chars().take(500).collect::<String>();
        if preview.trim().is_empty() {
            format!("{context} returned {status} with an empty body")
        } else {
            format!("{context} returned {status} with invalid JSON: {error}. Body: {preview}")
        }
    })
}

async fn get_torrent(
    client: &reqwest::Client,
    token: &str,
    torrent_id: i64,
) -> Result<serde_json::Value, String> {
    let response = client
        .get("https://api.torbox.app/v1/api/torrents/mylist")
        .bearer_auth(token)
        .query(&[
            ("id", torrent_id.to_string()),
            ("bypass_cache", "true".to_string()),
        ])
        .send()
        .await
        .map_err(|error| error.to_string())?;

    let body = json_response(response, "Torbox torrent details").await?;

    body.get("data").cloned().ok_or_else(|| {
        body_detail(&body).unwrap_or_else(|| "Torbox torrent details were unavailable.".to_string())
    })
}

fn choose_file_id(
    torrent: &serde_json::Value,
    file_idx: Option<i64>,
    filename: Option<&str>,
) -> Option<i64> {
    let files = torrent.get("files")?.as_array()?;

    if let Some(filename) = filename {
        let needle = filename.to_lowercase();
        if let Some(file) = files.iter().find(|file| {
            file.get("name")
                .or_else(|| file.get("short_name"))
                .and_then(|value| value.as_str())
                .map(|name| {
                    needle.contains(&name.to_lowercase()) || name.to_lowercase().contains(&needle)
                })
                .unwrap_or(false)
        }) {
            return numeric_id(file);
        }
    }

    if let Some(index) = file_idx.and_then(|value| usize::try_from(value).ok()) {
        if let Some(file) = files.get(index) {
            if let Some(id) = numeric_id(file) {
                return Some(id);
            }
        }
    }

    files
        .iter()
        .filter(|file| {
            file.get("name")
                .or_else(|| file.get("short_name"))
                .and_then(|value| value.as_str())
                .map(is_webkit_playable_filename)
                .unwrap_or(false)
        })
        .max_by_key(|file| {
            file.get("size")
                .or_else(|| file.get("bytes"))
                .and_then(json_to_i64)
                .unwrap_or(0)
        })
        .and_then(numeric_id)
        .or_else(|| {
            files
                .iter()
                .filter(|file| {
                    file.get("name")
                        .or_else(|| file.get("short_name"))
                        .and_then(|value| value.as_str())
                        .map(is_video_filename)
                        .unwrap_or(false)
                })
                .max_by_key(|file| {
                    file.get("size")
                        .or_else(|| file.get("bytes"))
                        .and_then(json_to_i64)
                        .unwrap_or(0)
                })
                .and_then(numeric_id)
        })
        .or_else(|| files.first().and_then(numeric_id))
}

async fn request_download_link(
    client: &reqwest::Client,
    token: &str,
    torrent_id: i64,
    file_id: i64,
) -> Result<String, String> {
    let response = client
        .get("https://api.torbox.app/v1/api/torrents/requestdl")
        .query(&[
            ("token", token.to_string()),
            ("torrent_id", torrent_id.to_string()),
            ("file_id", file_id.to_string()),
            ("zip_link", "false".to_string()),
            ("redirect", "false".to_string()),
            ("append_name", "false".to_string()),
        ])
        .send()
        .await
        .map_err(|error| error.to_string())?;

    let status = response.status();
    let body = response
        .json::<serde_json::Value>()
        .await
        .map_err(|error| error.to_string())?;

    if !status.is_success() || body.get("success").and_then(|value| value.as_bool()) == Some(false)
    {
        return Err(body_detail(&body)
            .unwrap_or_else(|| format!("Torbox download request failed with {status}")));
    }

    extract_download_url(&body)
        .ok_or_else(|| format!("Torbox did not return a playable download URL: {body}"))
}

fn extract_download_url(body: &serde_json::Value) -> Option<String> {
    [
        body.get("data"),
        body.get("download_url"),
        body.get("downloadUrl"),
        body.get("url"),
        body.get("link"),
    ]
    .into_iter()
    .flatten()
    .find_map(extract_download_url_from_value)
}

fn extract_download_url_from_value(value: &serde_json::Value) -> Option<String> {
    if let Some(url) = value
        .as_str()
        .filter(|url| url.starts_with("http://") || url.starts_with("https://"))
    {
        return Some(url.to_string());
    }

    [
        "download_url",
        "downloadUrl",
        "url",
        "link",
        "direct_link",
        "directLink",
    ]
    .into_iter()
    .find_map(|key| {
        value
            .get(key)
            .and_then(|child| child.as_str())
            .filter(|url| url.starts_with("http://") || url.starts_with("https://"))
            .map(ToString::to_string)
    })
}

fn numeric_id(value: &serde_json::Value) -> Option<i64> {
    value
        .get("id")
        .or_else(|| value.get("file_id"))
        .or_else(|| value.get("torrent_id"))
        .and_then(json_to_i64)
}

fn json_to_i64(value: &serde_json::Value) -> Option<i64> {
    value
        .as_i64()
        .or_else(|| value.as_u64().and_then(|number| i64::try_from(number).ok()))
        .or_else(|| value.as_str().and_then(|text| text.parse::<i64>().ok()))
}

fn is_video_filename(name: &str) -> bool {
    let lower = name.to_lowercase();
    [".mp4", ".mkv", ".webm", ".mov", ".m4v", ".avi"]
        .iter()
        .any(|extension| lower.ends_with(extension))
}

fn is_webkit_playable_filename(name: &str) -> bool {
    let lower = name.to_lowercase();
    [".mp4", ".m4v", ".mov", ".m3u8"]
        .iter()
        .any(|extension| lower.ends_with(extension))
}

fn body_detail(body: &serde_json::Value) -> Option<String> {
    body.get("detail")
        .or_else(|| body.get("error"))
        .or_else(|| body.get("message"))
        .or_else(|| body.get("msg"))
        .and_then(|value| value.as_str())
        .map(ToString::to_string)
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct TorboxAccountSummary {
    email: Option<String>,
    plan: Option<String>,
    plan_id: Option<i64>,
    premium: bool,
    expires_at: Option<String>,
    total_torrents: usize,
    active_torrents: usize,
    cached_torrents: usize,
}

fn plan_label(plan_id: i64) -> String {
    match plan_id {
        0 => "Free".to_string(),
        1 => "Essential".to_string(),
        2 => "Pro".to_string(),
        3 => "Standard".to_string(),
        _ => format!("Plan {plan_id}"),
    }
}

fn string_field(body: &serde_json::Value, keys: &[&str]) -> Option<String> {
    keys.iter().find_map(|key| {
        body.get(*key)
            .and_then(|value| value.as_str())
            .map(ToString::to_string)
    })
}

fn number_field(body: &serde_json::Value, keys: &[&str]) -> Option<i64> {
    keys.iter()
        .find_map(|key| body.get(*key).and_then(|value| value.as_i64()))
}

#[tauri::command]
pub(crate) async fn get_torbox_account(api_key: String) -> Result<TorboxAccountSummary, String> {
    let token = api_key.trim();
    if token.is_empty() {
        return Err("Add your Torbox API key to load account details.".into());
    }

    let client = reqwest::Client::new();
    let user_response = client
        .get("https://api.torbox.app/v1/api/user/me")
        .header("Authorization", format!("Bearer {token}"))
        .send()
        .await
        .map_err(|error| error.to_string())?;
    let user_status = user_response.status();
    let user_body = user_response
        .json::<serde_json::Value>()
        .await
        .map_err(|error| error.to_string())?;
    if !user_status.is_success() {
        return Err(body_detail(&user_body)
            .unwrap_or_else(|| format!("Torbox account lookup failed with {user_status}")));
    }

    let user = user_body.get("data").unwrap_or(&user_body);
    let plan_id = number_field(user, &["plan", "plan_id", "planId"]);
    let premium = user
        .get("premium")
        .or_else(|| user.get("is_premium"))
        .and_then(|value| value.as_bool())
        .unwrap_or(plan_id.unwrap_or(0) > 0);

    let torrent_response = client
        .get("https://api.torbox.app/v1/api/torrents/mylist")
        .header("Authorization", format!("Bearer {token}"))
        .send()
        .await
        .map_err(|error| error.to_string())?;
    let torrent_body = torrent_response
        .json::<serde_json::Value>()
        .await
        .unwrap_or(serde_json::json!([]));
    let torrents = torrent_body
        .get("data")
        .and_then(|value| value.as_array())
        .cloned()
        .or_else(|| torrent_body.as_array().cloned())
        .unwrap_or_default();

    let mut active_torrents = 0usize;
    let mut cached_torrents = 0usize;
    for entry in &torrents {
        let state = entry
            .get("download_state")
            .or_else(|| entry.get("state"))
            .and_then(|value| value.as_str())
            .unwrap_or("")
            .to_ascii_lowercase();
        if state.contains("cached")
            || entry.get("cached").and_then(|value| value.as_bool()) == Some(true)
        {
            cached_torrents += 1;
        }
        if !state.is_empty()
            && !["completed", "complete", "cached", "error"]
                .iter()
                .any(|item| state.contains(item))
        {
            active_torrents += 1;
        }
    }

    Ok(TorboxAccountSummary {
        email: string_field(user, &["email", "user_email"]),
        plan: string_field(user, &["plan_name", "planName"]).or(plan_id.map(plan_label)),
        plan_id,
        premium,
        expires_at: string_field(user, &["premium_expires_at", "expires_at", "expiration"]),
        total_torrents: torrents.len(),
        active_torrents,
        cached_torrents,
    })
}
